account.go

```diff
@@ -0,0 +1,356 @@
+		currentBlockNumber, err = account.client.CurrentBlockNumber(ctx)
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-time.After(5 * time.Millisecond):
+			}
+			continue
+		}
+	}
+	return nil
+}
+
+// Transfer transfers eth from the account to an ethereum address.
+func (account *account) Transfer(ctx context.Context, to common.Address, value *big.Int, confirmBlocks int64) error {
+
+	// Pre-condition check: Check if the account has enough balance
+	preConditionCheck := func() bool {
+		accountBalance, err := account.client.BalanceOf(ctx, account.Address(), &account.callOpts)
+		return err == nil && accountBalance.Cmp(value) >= 0
+	}
+
+	// Transaction: Transfer eth to address
+	f := func(transactOpts bind.TransactOpts) (*types.Transaction, error) {
+		bound := bind.NewBoundContract(to, abi.ABI{}, nil, account.client.EthClient, nil)
+
+		transactor := &bind.TransactOpts{
+			From:     transactOpts.From,
+			Nonce:    transactOpts.Nonce,
+			Signer:   transactOpts.Signer,
+			Value:    value,
+			GasPrice: transactOpts.GasPrice,
+			GasLimit: 21000,
+			Context:  transactOpts.Context,
+		}
+
+		return bound.Transfer(transactor)
+	}
+
+	return account.Transact(ctx, preConditionCheck, f, nil, confirmBlocks)
+}
+
+// retryNonceTx retries transaction execution on the blockchain until nonce
+// errors are not seen, or until the context times out.
+func (account *account) retryNonceTx(ctx context.Context, f func(bind.TransactOpts) (*types.Transaction, error)) (*types.Transaction, error) {
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	tx, err := f(account.transactOpts)
+
+	// On successful execution, increment nonce in transactOpts and return
+	if err == nil {
+		account.transactOpts.Nonce.Add(account.transactOpts.Nonce, big.NewInt(1))
+		return tx, nil
+	}
+
+	// Process errors to check for nonce issues
+	// If error indicates that nonce is too low, increment nonce and retry
+	if err == core.ErrNonceTooLow || strings.Contains(err.Error(), "nonce is too low") {
+		account.transactOpts.Nonce.Add(account.transactOpts.Nonce, big.NewInt(1))
+		return account.retryNonceTx(ctx, f)
+	}
+
+	// If error indicates that nonce is too low, decrement nonce and retry
+	if err == core.ErrNonceTooHigh || strings.Contains(err.Error(), "nonce is too high") {
+		account.transactOpts.Nonce.Sub(account.transactOpts.Nonce, big.NewInt(1))
+		return account.retryNonceTx(ctx, f)
+	}
+
+	// If any other type of nonce error occurs we will refresh the nonce and
+	// try again for up to 1 minute
+	var nonce uint64
+	for try := 0; try < 60 && strings.Contains(err.Error(), "nonce"); try++ {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(time.Second):
+		}
+
+		// Get updated nonce and retry 'f'
+		nonce, err = account.client.EthClient.PendingNonceAt(ctx, account.transactOpts.From)
+		if err != nil {
+			continue
+		}
+		account.transactOpts.Nonce = big.NewInt(int64(nonce))
```

Comment:
Do we need to lock the mutex when modifying the nonce?

Reply:
`retryNonceTx` is a private method that is called by `Transact` after locking the mutex.